File: 023_Merge_k_Sorted_Lists/023_1.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def mergeKLists(self, lists):
        length = len(lists)
        interval = 1
        while interval < length:
            for i in range(0, length - interval, interval * 2):
                lists[i] = self.merge2Lists(lists[i], lists[i + interval])
            interval *= 2
        return lists[0] if length > 0 else lists

    def merge2Lists(self, l1, l2):
        dummy = cur = ListNode(0)
        while l1 and l2:
            if l1.val <= l2.val:
                cur.next = l1
                l1 = l1.next
            else:
                cur.next = l2
                l2 = l2.next
            cur = cur.next
        if not l1:
            cur.next = l2
        else:
            cur.next = l1
        return dummy.next


if __name__ == "__main__":
    l1 = ListNode(1)
    l2 = ListNode(5)
    l3 = ListNode(7)
    l1.next = l2
    l2.next = l3
    l4 = ListNode(2)
    l5 = ListNode(4)
    l6 = ListNode(6)
    l4.next = l5
    l5.next = l6
    l7 = ListNode(3)
    l8 = ListNode(8)
    l9 = ListNode(9)
    l7.next = l8
    l8.next = l9
    s = Solution()
    res = s.mergeKLists([l1, l4, l7])
    while res:
        print(res.val)
        res = res.next

File: 023_Merge_k_Sorted_Lists/023_1.cpp
struct ListNode {
	int val;
	ListNode *next;
	ListNode(int x): val(x), next(NULL) { };
};


class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        size_t listLength = lists.size();
        if (listLength == 0)  return NULL;
        int interval = 1;
        while (interval < listLength) {
            for (int i = 0; i < listLength - interval; i += (interval * 2))
                lists[i] = mergeTwoLists(lists[i], lists[i + interval]);
            interval *= 2;
        }
        return lists[0];
    }

    ListNode * mergeTwoLists(ListNode *l1, ListNode *l2) {
                ListNode *dummy, *cur;
        dummy = cur = new ListNode(INT32_MIN);
        while (l1 && l2) {
            if (l1->val <= l2->val) {
                cur->next = l1;
                l1 = l1->next;
            } else {
                cur->next = l2;
                l2 = l2->next;
            }
            cur = cur->next;
        }
        if (!l1)
            cur->next = l2;
        if (!l2)
            cur->next = l1;
        return dummy->next;
    }
};